### ISSUE TITLE
[PriorityQueue] Custom comparator for priorities

### ISSUE THREAD

**Issue by enumag** [php-ds/extension#121]
I'd like to use the PriorityQueue but in my case the values I need to use for ordering are dates and I need the oldest one first. This means I can't use PriorityQueue because the order would be reversed.

In my opinion PriorityQueue should accept an optional comparator callback to decide the order.

**Comment by rtheunissen** [comment]
Could you use `-timestamp` for the priority?

**Comment by enumag** [comment]
I think the order would still be oposite to what I need with timestamp.

**Comment by enumag** [comment]
Negative timestamp huh? Yeah, I guess that could work. It feels like a dirty workaround rather than proper solution though.

**Comment by rtheunissen** [comment]
> It feels like a dirty workaround rather than proper solution though.

I agree. 2.0 will have a Heap with an optional comparator that would work here.

**Comment by enumag** [comment]
Would it be possible to have another similar structure with custom comparator and non-destructive iteration?

**Comment by rtheunissen** [comment]
Like a binary search tree? There is one on the way. :+1:
